R/utils.R

```diff
@@ -1,91 +1,137 @@
+
+    if (sum(structure_vec) != (cols - 1)) {
+      cli_abort(c(
+        "Sum of structure vector must equal the number of columns in matrix",
+        " minus 1"
+      ))
     }
   }
 
+  # For each row, determine which columns should be NA
+  cutoff_cols <- cumsum(structure_vec) + 1
+
+  for (i in seq_along(structure_vec)) {
+    index_row <- rows - i + 1
+    start_col <- cutoff_cols[i]
+    result[index_row, start_col:cols] <- NA_real_
+  }
+
   return(result)
 }
 
-#' Check if matrix only contains NAs in the bottom right
+.expand_structure_vec <- function(structure, cols) {
+  if (structure <= 0) {
+    cli_abort("Structure must be positive")
+  }
+
+  if (structure > cols) {
+    cli_abort("Structure cannot be larger than number of columns")
+  }
+  adjusted_cols <- cols - 1
+
+  n_complete_groups <- floor(adjusted_cols / structure)
+
+  structure_vec <- rep(structure, n_complete_groups)
+
+  remainder <- adjusted_cols %% structure
+
+  if (remainder > 0) {
+    structure_vec <- c(structure_vec, remainder)
+  }
+
+  return(structure_vec)
+}
+
+#' Check if matrix has valid NA pattern
 #'
 #' @param mat Matrix
 #'
 #' @returns Boolean indicating whether the matrix only contains NAs in the
 #'    bottom right (TRUE if only in bottom right, FALSE if elsewhere).
 #' @keywords internal
 .check_na_bottom_right <- function(mat) {
-  # Create a logical mask for valid NA positions
   n_rows <- nrow(mat)
-  mask <- matrix(FALSE, nrow = n_rows, ncol = ncol(mat))
+  n_cols <- ncol(mat)
 
-  for (i in seq_len(n_rows)) {
-    cutoff <- n_rows - i + 1
-    if (cutoff < ncol(mat)) {
-      mask[i, (cutoff + 1):ncol(mat)] <- TRUE
+  for (i in 1:n_rows) {
+    row_data <- mat[i, ]
+    na_indices <- which(is.na(row_data))
+
+    # If there are NAs in this row
+    if (length(na_indices) > 0) {
+      min_na_idx <- min(na_indices)
+      # Check that all entries from the first NA onwards are also NA
+      if (!all(is.na(row_data[min_na_idx:n_cols]))) {
+        return(FALSE)
+      }
     }
   }
 
-  # Check if any NAs exist outside valid region
-  invalid_nas <- sum(is.na(mat) & !mask)
-  return(invalid_nas == 0)
-}
+  # Check column consistency (if a cell is NA, all cells below it must be NA)
+  for (j in 1:n_cols) {
+    col_data <- mat[, j]
+    na_indices <- which(is.na(col_data))
 
-#' Check if matrix only contains zeros in the bottom right
-#'
-#' @param mat Matrix
-#'
-#' @returns Boolean indicating whether the matrix only contains zeros in the
-#'    bottom right (if TRUE, entire bottom right is 0s)
-#' @keywords internal
-.check_zeros_bottom_right <- function(mat) {
-  n_rows <- nrow(mat)
-  mask <- matrix(FALSE, nrow = n_rows, ncol = ncol(mat))
-
-  for (i in seq_len(n_rows)) {
-    cutoff <- n_rows - i + 1
-    if (cutoff < ncol(mat)) {
-      mask[i, (cutoff + 1):ncol(mat)] <- TRUE
+    if (length(na_indices) > 0) {
+      min_na_idx <- min(na_indices)
+      if (!all(is.na(col_data[min_na_idx:n_rows]))) {
+        return(FALSE)
```

Comment:
Looks like we should add tests to check column consistency

Reply:
I think it was just that this code hadn't been tested but I see that these tests have been added.

Reply:
(and perhaps were there and I missed them)